File: DirectShowFilters/TsWriter/source/NttParser.cpp
/* 
 *	Copyright (C) 2006-2008 Team MediaPortal
 *	http://www.team-mediaportal.com
 *
 *  This Program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *   
 *  This Program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *   
 *  You should have received a copy of the GNU General Public License
 *  along with GNU Make; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA. 
 *  http://www.gnu.org/copyleft/gpl.html
 *
 */
#include "NttParser.h"

extern void LogDebug(const char *fmt, ...);
extern bool DisableCRCCheck();

CNttParser::CNttParser(void)
{
  //SetPid(0);
  if (DisableCRCCheck())
  {
    EnableCrcCheck(false);
  }
  Reset();
  m_pCallBack = NULL;
}

CNttParser::~CNttParser(void)
{
}

void CNttParser::Reset()
{
  LogDebug("NttParser: reset");
  CSectionDecoder::Reset();
  LogDebug("NttParser: reset done");
}

void CNttParser::SetCallBack(INttCallBack* callBack)
{
  m_pCallBack = callBack;
}

void CNttParser::OnNewSection(CSection& sections)
{
  if (sections.table_id != 0xc3)
  {
    LogDebug("NttParser: other table ID 0x%x", sections.table_id);
    return;
  }
  if (m_pCallBack == NULL)
  {
    return;
  }
  byte* section = sections.Data;

  try
  {
    int section_length = ((section[1] & 0x0f) << 8) + section[2];
    if (section_length > 1021 || section_length < 9)
    {
      LogDebug("NttParser: invalid section length = %d", section_length);
      return;
    }
    int protocol_version = (section[3] & 0x1f);
    unsigned int iso_639_language_code = (section[4] << 16) + (section[5] << 8) + section[6];
    int transmission_medium = (section[7] >> 4);
    int table_subtype = (section[7] & 0xf);

    LogDebug("NttParser: section length = %d, protocol version = %d, ISO 639 language = %s, transmission medium = %d, table subtype = %d", section_length, protocol_version, &iso_639_language_code, transmission_medium, table_subtype);
    if (table_subtype != 6)
    {
      LogDebug("NttParser: unsupported table subtype %d", table_subtype);
      return;
    }

    //-------------------------------------------------------------------------
    // Source name subtable...
    int number_of_sns_records = section[8];
    LogDebug("NttParser: number of SNS records = %d", number_of_sns_records);
    int pointer = 9;
    int endOfSection = section_length - 1;
    for (int i = 0; i < number_of_sns_records; i++)
    {
      if (pointer + 4 > endOfSection)
      {
        LogDebug("NttParser: detected number of SNS records %d is invalid in loop %d, pointer = %d, end of section = %d, section length = %d", number_of_sns_records, i, pointer, endOfSection, section_length);
        return;
      }

      int application_type = (section[pointer++] >> 7);
      int source_id = (section[pointer] << 8) + section[pointer + 1];   // source_id when application_type is 0, otherwise application_id
      pointer += 2;
      int name_length = section[pointer++];
      LogDebug("NttParser: source ID = 0x%x, application type = %d, name length = %d", source_id, application_type, name_length);

      if (pointer + name_length + 1 > endOfSection)
      {
        LogDebug("NttParser: invalid name length = %d, pointer = %d, end of section = %d, section length = %d", name_length, pointer, endOfSection, section_length);
        return;
      }

      char* name = NULL;
      DecodeMultilingualText(&section[pointer], name_length, &name);
      if (name != NULL)
      {
        LogDebug("NttParser: name = %s", name);
        if (m_pCallBack != NULL)
        {
          m_pCallBack->OnNttReceived(source_id, application_type, name, iso_639_language_code);
        }
        else
        {
          delete[] name;
          name = NULL;
        }
      }
      pointer += name_length;

      int sns_descriptors_count = section[pointer++];
      LogDebug("NttParser: SNS decriptor count = %d", sns_descriptors_count);
      for (int j = 0; j < sns_descriptors_count; j++)
      {
        if (pointer + 1 > endOfSection)
        {
          LogDebug("NttParser: detected SNS descriptor count %d is invalid in loop %d, pointer = %d, end of section = %d", sns_descriptors_count, j, pointer, endOfSection);
          return;
        }

        int tag = section[pointer++];
        int length = section[pointer++];
        LogDebug("NttParser: SNS descriptor, tag = 0x%x, length = %d, pointer = %d, end of descriptor = %d, end of section = %d, section length = %d", tag, length, pointer, pointer + length, endOfSection, section_length);
        if (pointer + length > endOfSection)
        {
          LogDebug("NttParser: invalid SNS descriptor length = %d, pointer = %d, end of descriptor = %d, end of section = %d, section length = %d", length, pointer, pointer + length, endOfSection, section_length);
          return;
        }

        pointer += length;
      }
    }
    //-------------------------------------------------------------------------

    while (pointer + 1 < endOfSection)
    {
      int tag = section[pointer++];
      int length = section[pointer++];
      LogDebug("NttParser: descriptor, tag = 0x%x, length = %d, pointer = %d, end of descriptor = %d, end of section = %d, section length = %d", tag, length, pointer, pointer + length, endOfSection, section_length);
      if (pointer + length > endOfSection)
      {
        LogDebug("NttParser: invalid descriptor length = %d, pointer = %d, end of descriptor = %d, end of section = %d, section length = %d", length, pointer, pointer + length, endOfSection, section_length);
        return;
      }

      pointer += length;
    }

    if (pointer != endOfSection)
    {
      LogDebug("NttParser: section parsing error");
    }
  }
  catch (...)
  {
    LogDebug("NttParser: unhandled exception in OnNewSection()");
  }
}

void CNttParser::DecodeMultilingualText(byte* b, int length, char** string)
{
  if (length < 2)
  {
    LogDebug("NttParser: invalid multilingual text length = %d", length);
    return;
  }
  *string = new char[length + 1];
  if (*string == NULL)
  {
    LogDebug("NttParser: failed to allocate memory in DecodeMultilingualText()");
    return;
  }
  int stringOffset = 0;

  int pointer = 0;
  while (pointer + 1 < length)
  {
    int mode = b[pointer++];
    int segment_length = b[pointer++];
    if (pointer + segment_length > length)
    {
      LogDebug("NttParser: invalid multilingual text segment length = %d, pointer = %d, string offset = %d, mode = 0x%x, text length = %d", segment_length, pointer, stringOffset, mode, length);
      delete[] *string;
      *string = NULL;
      return;
    }

    if (mode == 0)
    {
      // We only support ASCII encoding at this time.
      memcpy(*string, &b[pointer], segment_length);
      stringOffset += segment_length;
      pointer += segment_length;
      (*string)[stringOffset] = 0;  // NULL terminate
    }
    else
    {
      LogDebug("NttParser: unsupported segment mode in DecodeMultilingualText(), mode = 0x%x", mode);
      for (int i = 0; i < segment_length; i++)
      {
        LogDebug("  %d: 0x%x", b[pointer++]);
      }
    }
  }
}

File: TvEngine3/TVLibrary/TVLibrary/Implementations/Dri/Parser/NttParser.cs
#region Copyright (C) 2005-2011 Team MediaPortal

// Copyright (C) 2005-2011 Team MediaPortal
// http://www.team-mediaportal.com
// 
// MediaPortal is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
// 
// MediaPortal is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MediaPortal. If not, see <http://www.gnu.org/licenses/>.

#endregion

using System;
using TvLibrary.Implementations.DVB;

namespace TvLibrary.Implementations.Dri.Parser
{
  public class NttParser
  {
    private enum TableSubtype
    {
      // ATSC A-56 seems to be out of date
      TransponderName = 1,
      SatelliteText,
      RatingsText,
      RatingSystem,
      CurrencySystem,
      SourceName,
      MapName
    }

    public void Decode(byte[] section)
    {
      if (section == null || section.Length < 14)
      {
        Log.Log.Error("NTT: invalid section length");
        return;
      }

      byte tableId = section[2];
      if (tableId != 0xc3)
      {
        return;
      }
      int sectionLength = ((section[3] & 0x3f) << 8) + section[4];  // difference between A-56 and SCTE 65
      if (section.Length != 2 + sectionLength + 3)
      {
        Log.Log.Error("NTT: invalid section length = {0}, byte count = {1}", sectionLength, section.Length);
        return;
      }
      int protocolVersion = (section[5] & 0x1f);
      string isoLangCode = System.Text.Encoding.ASCII.GetString(section, 6, 3);
      AtscTransmissionMedium transmissionMedium = (AtscTransmissionMedium)(section[9] >> 4);
      TableSubtype tableType = (TableSubtype)(section[9] & 0x0f);
      Log.Log.Debug("NTT: protocol version = {0}, ISO language code = {1}, transmission medium = {2}, table type = {3}",
        protocolVersion, isoLangCode, transmissionMedium, tableType);

      int pointer = 10;
      int endOfSection = section.Length - 4;

      try
      {
        if (tableType == TableSubtype.TransponderName)
        {
          DecodeTransponderName(section, endOfSection, ref pointer);
        }
        else if (tableType == TableSubtype.SatelliteText)
        {
          DecodeSatelliteText(section, endOfSection, ref pointer);
        }
        else if (tableType == TableSubtype.RatingsText)
        {
          DecodeRatingsText(section, endOfSection, ref pointer);
        }
        else if (tableType == TableSubtype.RatingSystem)
        {
          DecodeRatingSystem(section, endOfSection, ref pointer);
        }
        else if (tableType == TableSubtype.CurrencySystem)
        {
          DecodeCurrencySystem(section, endOfSection, ref pointer);
        }
        else if (tableType == TableSubtype.SourceName)
        {
          DecodeSourceName(section, endOfSection, ref pointer);
        }
        else if (tableType == TableSubtype.MapName)
        {
          DecodeMapName(section, endOfSection, ref pointer);
        }
        else
        {
          Log.Log.Error("NTT: unsupported subtable type {0}", tableType);
          return;
        }
      }
      catch (Exception ex)
      {
        Log.Log.Error(ex.Message);
        return;
      }

      while (pointer + 1 < endOfSection)
      {
        byte tag = section[pointer++];
        byte length = section[pointer++];
        Log.Log.Debug("NTT: descriptor, tag = 0x{0:x}, length = {1}", tag, length);
        if (pointer + length > endOfSection)
        {
          Log.Log.Error("NTT: invalid descriptor length, pointer = {0}, end of section = {1}, descriptor length = {2}", pointer, endOfSection, length);
          return;
        }

        if (tag == 0x93)  // revision detection descriptor
        {
          ParserCommon.DecodeRevisionDetectionDescriptor(section, pointer, length);
        }

        pointer += length;
      }

      if (pointer != endOfSection)
      {
        Log.Log.Error("NTT: corruption detected at end of section, pointer = {0}, end of section = {1}", pointer, endOfSection);
        return;
      }
    }

    private void DecodeTransponderName(byte[] section, int endOfSection, ref int pointer)
    {
      if (pointer + 3 > endOfSection)
      {
        throw new Exception(string.Format("NTT: corruption detected at transponder name, pointer = {0}, end of section = {1}", pointer, endOfSection));
      }

      byte satelliteId = section[pointer++];
      byte firstIndex = section[pointer++];
      byte numberOfTntRecords = section[pointer++];
      Log.Log.Debug("NTT: transponder name, satellite ID = 0x{0:x}, first index = {1}, number of TNT records = {2}", satelliteId, firstIndex, numberOfTntRecords);

      for (byte i = 0; i < numberOfTntRecords; i++)
      {
        if (pointer + 3 > endOfSection)
        {
          throw new Exception(string.Format("NTT: detected transponder name subtable number of TNT records {0} is invalid in loop {1}, pointer = {2}, end of section = {3}", numberOfTntRecords, i, pointer, endOfSection));
        }
        int transponderNumber = (section[pointer++] & 0x3f);
        int transponderNameLength = (section[pointer++] & 0x1f);
        if (pointer + transponderNameLength > endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid transponder name subtable transponder name length {0}, pointer = {1}, end of section = {2}", transponderNameLength, pointer, endOfSection));
        }
        string transponderName = DecodeMultilingualText(section, pointer + transponderNameLength, ref pointer);
        Log.Log.Debug("NTT: transponder name, number = {0}, name = {1}", transponderNumber, transponderName);

        // subtable descriptors
        if (pointer >= endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid section length at transponder name subtable descriptor count, pointer = {0}, end of section = {1}", pointer, endOfSection));
        }
        byte descriptorCount = section[pointer++];
        for (byte d = 0; d < descriptorCount; d++)
        {
          if (pointer + 2 > endOfSection)
          {
            throw new Exception(string.Format("NTT: detected transponder name subtable descriptor count {0} is invalid in loop {1}, pointer = {2}, end of section = {3}", descriptorCount, d, pointer, endOfSection));
          }
          byte tag = section[pointer++];
          byte length = section[pointer++];
          Log.Log.Debug("NTT: transponder name subtable descriptor, tag = 0x{0:x}, length = {1}", tag, length);
          if (pointer + length > endOfSection)
          {
            throw new Exception(string.Format("NTT: invalid transponder name subtable descriptor length {0}, pointer = {1}, end of section = {2}", length, pointer, endOfSection));
          }
          pointer += length;
        }
      }
    }

    private void DecodeSatelliteText(byte[] section, int endOfSection, ref int pointer)
    {
      if (pointer + 2 > endOfSection)
      {
        throw new Exception(string.Format("NTT: corruption detected at satellite text, pointer = {0}, end of section = {1}", pointer, endOfSection));
      }

      byte firstIndex = section[pointer++];
      byte numberOfSttRecords = section[pointer++];
      Log.Log.Debug("NTT: satellite text, first index = {0}, number of STT records = {1}", firstIndex, numberOfSttRecords);

      for (byte i = 0; i < numberOfSttRecords; i++)
      {
        if (pointer + 4 > endOfSection)
        {
          throw new Exception(string.Format("NTT: detected satellite text subtable number of STT records {0} is invalid in loop {1}, pointer = {2}, end of section = {3}", numberOfSttRecords, i, pointer, endOfSection));
        }
        byte satelliteId = section[pointer++];
        int satelliteReferenceNameLength = (section[pointer++] & 0x0f);
        if (pointer + satelliteReferenceNameLength > endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid satellite text subtable satellite reference name length {0}, pointer = {1}, end of section = {2}", satelliteReferenceNameLength, pointer, endOfSection));
        }
        string satelliteReferenceName = DecodeMultilingualText(section, pointer + satelliteReferenceNameLength, ref pointer);

        if (pointer >= endOfSection)
        {
          throw new Exception(string.Format("NTT: corruption detected at satellite text subtable full satellite name, pointer = {0}, end of section = {1}", pointer, endOfSection));
        }
        int fullSatelliteNameLength = (section[pointer++] & 0x1f);
        if (pointer + fullSatelliteNameLength > endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid satellite text subtable full satellite name length {0}, pointer = {1}, end of section = {2}", fullSatelliteNameLength, pointer, endOfSection));
        }
        string fullSatelliteName = DecodeMultilingualText(section, pointer + fullSatelliteNameLength, ref pointer);

        Log.Log.Debug("NTT: satellite text, satellite ID = 0x{0:x}, reference name = {1}, full name = {2}", satelliteId, satelliteReferenceName, fullSatelliteName);

        // subtable descriptors
        if (pointer >= endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid section length at satellite text subtable descriptor count, pointer = {0}, end of section = {1}", pointer, endOfSection));
        }
        byte descriptorCount = section[pointer++];
        for (byte d = 0; d < descriptorCount; d++)
        {
          if (pointer + 2 > endOfSection)
          {
            throw new Exception(string.Format("NTT: detected satellite text subtable descriptor count {0} is invalid in loop {1}, pointer = {2}, end of section = {3}", descriptorCount, d, pointer, endOfSection));
          }
          byte tag = section[pointer++];
          byte length = section[pointer++];
          Log.Log.Debug("NTT: satellite text subtable descriptor, tag = 0x{0:x}, length = {1}", tag, length);
          if (pointer + length > endOfSection)
          {
            throw new Exception(string.Format("NTT: invalid satellite text subtable descriptor length {0}, pointer = {1}, end of section = {2}", length, pointer, endOfSection));
          }
          pointer += length;
        }
      }
    }

    private void DecodeRatingsText(byte[] section, int endOfSection, ref int pointer)
    {
      if (pointer >= endOfSection)
      {
        throw new Exception(string.Format("NTT: corruption detected at ratings text, pointer = {0}, end of section = {1}", pointer, endOfSection));
      }

      byte ratingRegion = section[pointer++];
      for (byte i = 0; i < 6; i++)
      {
        if (pointer >= endOfSection)
        {
          throw new Exception(string.Format("NTT: corruption detected at ratings text subtable levels defined loop {0}, pointer = {1}, end of section = {2}", i, pointer, endOfSection));
        }
        byte levelsDefined = section[pointer++];
        if (levelsDefined > 0)
        {
          if (pointer >= endOfSection)
          {
            throw new Exception(string.Format("NTT: corruption detected at ratings text subtable dimension name length, pointer = {0}, end of section = {1}", pointer, endOfSection));
          }
          byte dimensionNameLength = section[pointer++];
          if (pointer + dimensionNameLength > endOfSection)
          {
            throw new Exception(string.Format("NTT: invalid ratings text subtable dimension name length {0}, pointer = {1}, end of section = {2}", dimensionNameLength, pointer, endOfSection));
          }
          string dimensionName = DecodeMultilingualText(section, pointer + dimensionNameLength, ref pointer);
          Log.Log.Debug("NTT: ratings text, dimension name = {0}, levels defined = {1}", dimensionName, levelsDefined);
          for (byte l = 0; l < levelsDefined; l++)
          {
            byte ratingNameLength = section[pointer++];
            if (pointer + ratingNameLength > endOfSection)
            {
              throw new Exception(string.Format("NTT: invalid ratings text subtable rating name length {0}, pointer = {1}, end of section = {2}", ratingNameLength, pointer, endOfSection));
            }
            string ratingName = DecodeMultilingualText(section, pointer + ratingNameLength, ref pointer);
            Log.Log.Debug("  rating name = {0}", ratingName);
          }
        }
      }
    }

    private void DecodeRatingSystem(byte[] section, int endOfSection, ref int pointer)
    {
      if (pointer >= endOfSection)
      {
        throw new Exception(string.Format("NTT: corruption detected at rating system, pointer = {0}, end of section = {1}", pointer, endOfSection));
      }

      byte regionsDefined = section[pointer++];
      for (byte i = 0; i < regionsDefined; i++)
      {
        if (pointer + 3 > endOfSection)
        {
          throw new Exception(string.Format("NTT: detected rating system subtable regions defined {0} is invalid in loop {1}, pointer = {2}, end of section = {3}", regionsDefined, i, pointer, endOfSection));
        }
        byte dataLength = section[pointer++];
        int endOfData = pointer + dataLength;
        if (endOfData > endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid rating system subtable data length {0}, pointer = {1}, end of section = {2}", dataLength, pointer, endOfSection));
        }
        byte ratingRegion = section[pointer++];
        byte stringLength = section[pointer++];
        if (pointer + stringLength > endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid rating system subtable string length {0}, pointer = {1}, end of section = {2}", stringLength, pointer, endOfSection));
        }
        string ratingSystem = DecodeMultilingualText(section, pointer + stringLength, ref pointer);
        Log.Log.Debug("NTT: rating system, region = {0}, system = {1}", ratingRegion, ratingSystem);

        // subtable descriptors
        while (pointer + 1 < endOfData)
        {
          byte tag = section[pointer++];
          byte length = section[pointer++];
          Log.Log.Debug("NTT: rating system subtable descriptor, tag = 0x{0:x}, length = {1}", tag, length);
          if (pointer + length > endOfSection)
          {
            throw new Exception(string.Format("NTT: invalid rating system subtable descriptor length {0}, pointer = {1}, end of section = {2}", length, pointer, endOfSection));
          }
          pointer += length;
        }
        if (pointer != endOfData)
        {
          throw new Exception(string.Format("NTT: corruption detected at end of rating system data, pointer = {0}, end of section = {1}, end of data = {2}", pointer, endOfSection, endOfData));
        }
      }
    }

    private void DecodeCurrencySystem(byte[] section, int endOfSection, ref int pointer)
    {
      if (pointer >= endOfSection)
      {
        throw new Exception(string.Format("NTT: corruption detected at currency system, pointer = {0}, end of section = {1}", pointer, endOfSection));
      }

      byte regionsDefined = section[pointer++];
      for (byte i = 0; i < regionsDefined; i++)
      {
        if (pointer + 3 > endOfSection)
        {
          throw new Exception(string.Format("NTT: detected currency system subtable regions defined {0} is invalid in loop {1}, pointer = {2}, end of section = {3}", regionsDefined, i, pointer, endOfSection));
        }
        byte dataLength = section[pointer++];
        int endOfData = pointer + dataLength;
        if (endOfData > endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid currency system subtable data length {0}, pointer = {1}, end of section = {2}", dataLength, pointer, endOfSection));
        }
        byte currencyRegion = section[pointer++];
        byte stringLength = section[pointer++];
        if (pointer + stringLength > endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid currency system subtable string length {0}, pointer = {1}, end of section = {2}", stringLength, pointer, endOfSection));
        }
        string currencySystem = DecodeMultilingualText(section, pointer + stringLength, ref pointer);
        Log.Log.Debug("NTT: currency system, region = {0}, system = {1}", currencyRegion, currencySystem);

        // subtable descriptors
        while (pointer + 1 < endOfData)
        {
          byte tag = section[pointer++];
          byte length = section[pointer++];
          Log.Log.Debug("NTT: currency system subtable descriptor, tag = 0x{0:x}, length = {1}", tag, length);
          if (pointer + length > endOfSection)
          {
            throw new Exception(string.Format("NTT: invalid currency system subtable descriptor length {0}, pointer = {1}, end of section = {2}", length, pointer, endOfSection));
          }
          pointer += length;
        }
        if (pointer != endOfData)
        {
          throw new Exception(string.Format("NTT: corruption detected at end of currency system data, pointer = {0}, end of section = {1}, end of data = {2}", pointer, endOfSection, endOfData));
        }
      }
    }

    private void DecodeSourceName(byte[] section, int endOfSection, ref int pointer)
    {
      if (pointer >= endOfSection)
      {
        throw new Exception(string.Format("NTT: corruption detected at source name, pointer = {0}, end of section = {1}", pointer, endOfSection));
      }

      byte numberOfSntRecords = section[pointer++];
      for (byte i = 0; i < numberOfSntRecords; i++)
      {
        if (pointer + 5 > endOfSection)
        {
          throw new Exception(string.Format("NTT: detected source name subtable number of SNT records {0} is invalid in loop {1}, pointer = {2}, end of section = {3}", numberOfSntRecords, i, pointer, endOfSection));
        }
        bool isApplication = ((section[pointer++] & 0x80) != 0);
        int sourceId = (section[pointer] << 8) + section[pointer + 1];
        pointer += 2;
        byte nameLength = section[pointer++];
        if (pointer + nameLength > endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid source name subtable string length {0}, pointer = {1}, end of section = {2}", nameLength, pointer, endOfSection));
        }
        string sourceName = DecodeMultilingualText(section, pointer + nameLength, ref pointer);
        Log.Log.Debug("NTT: source name, source ID = 0x{0:x}, name = {1}, is application = {2}", sourceId, sourceName, isApplication);

        // subtable descriptors
        if (pointer >= endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid section length at source name subtable descriptor count, pointer = {0}, end of section = {1}", pointer, endOfSection));
        }
        byte descriptorCount = section[pointer++];
        for (byte d = 0; d < descriptorCount; d++)
        {
          if (pointer + 2 > endOfSection)
          {
            throw new Exception(string.Format("NTT: detected source name subtable descriptor count {0} is invalid in loop {1}, pointer = {2}, end of section = {3}", descriptorCount, d, pointer, endOfSection));
          }
          byte tag = section[pointer++];
          byte length = section[pointer++];
          Log.Log.Debug("NTT: source name subtable descriptor, tag = 0x{0:x}, length = {1}", tag, length);
          if (pointer + length > endOfSection)
          {
            throw new Exception(string.Format("NTT: invalid source name subtable descriptor length {0}, pointer = {1}, end of section = {2}", length, pointer, endOfSection));
          }
          pointer += length;
        }
      }
    }

    private void DecodeMapName(byte[] section, int endOfSection, ref int pointer)
    {
      if (pointer >= endOfSection)
      {
        throw new Exception(string.Format("NTT: corruption detected at map name, pointer = {0}, end of section = {1}", pointer, endOfSection));
      }

      byte numberOfMntRecords = section[pointer++];
      for (byte i = 0; i < numberOfMntRecords; i++)
      {
        if (pointer + 4 > endOfSection)
        {
          throw new Exception(string.Format("NTT: detected map name subtable number of MNT records {0} is invalid in loop {1}, pointer = {2}, end of section = {3}", numberOfMntRecords, i, pointer, endOfSection));
        }
        int vctId = (section[pointer] << 8) + section[pointer + 1];
        pointer += 2;
        byte mapNameLength = section[pointer++];
        if (pointer + mapNameLength > endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid map name subtable map name length {0}, pointer = {1}, end of section = {2}", mapNameLength, pointer, endOfSection));
        }
        string mapName = DecodeMultilingualText(section, pointer + mapNameLength, ref pointer);
        Log.Log.Debug("NTT: map name, VCT ID = 0x{0:x}, name = {1}", vctId, mapName);

        // subtable descriptors
        if (pointer >= endOfSection)
        {
          throw new Exception(string.Format("NTT: invalid section length at map name subtable descriptor count, pointer = {0}, end of section = {1}", pointer, endOfSection));
        }
        byte descriptorCount = section[pointer++];
        for (byte d = 0; d < descriptorCount; d++)
        {
          if (pointer + 2 > endOfSection)
          {
            throw new Exception(string.Format("NTT: detected map name subtable descriptor count {0} is invalid in loop {1}, pointer = {2}, end of section = {3}", descriptorCount, d, pointer, endOfSection));
          }
          byte tag = section[pointer++];
          byte length = section[pointer++];
          Log.Log.Debug("NTT: map name subtable descriptor, tag = 0x{0:x}, length = {1}", tag, length);
          if (pointer + length > endOfSection)
          {
            throw new Exception(string.Format("NTT: invalid map name subtable descriptor length {0}, pointer = {1}, end of section = {2}", length, pointer, endOfSection));
          }
          pointer += length;
        }
      }
    }

    public string DecodeMultilingualText(byte[] section, int endOfString, ref int pointer)
    {
      string text = string.Empty;
      while (pointer + 1 < endOfString)
      {
        byte mode = section[pointer++];
        byte segmentLength = section[pointer++];
        if (pointer + segmentLength > endOfString)
        {
          throw new Exception(string.Format("NTT: invalid multilingual text segment length {0}, pointer = {1}, end of string = {2}", segmentLength, pointer, endOfString));
        }

        if (mode == 0)
        {
          // We only support ASCII encoding at this time.
          text += System.Text.Encoding.ASCII.GetString(section, pointer, segmentLength);
        }
        else
        {
          Log.Log.Debug("NTT: unsupported multilingual text segment mode 0x{0:x}", mode);
          DVB_MMI.DumpBinary(section, pointer, segmentLength);
        }
        pointer += segmentLength;
      }
      if (pointer != endOfString)
      {
        throw new Exception(string.Format("NTT: corruption detected at end of multilingual string, pointer = {0}, end of string = {2}", pointer, endOfString));
      }
      return text;
    }
  }
}
